### ISSUE TITLE
Caught between two use cases

### ISSUE THREAD

**Issue by karlnp** [wallabyjs/public#402]
Hi Artem. I have to say, kudos on your dedicated support of Wallaby - I've had so much trouble that ordinarily I'd begin to regret my purchase, but it's really gratifying to see you being proactive and helpful. That said:

I think I'm running up against a fundamental difference in how my project works vs wallaby's assumptions of the kinds of projects it would be used to test.

Here's how my project works standalone and with karma as a test runner:

* Webpack compiles and translates the application code, which is a mix of ES5 and ES2015 and ES7
	* Webpack injects build-contextual constants via the DefinePlugin, which essentially adds global string vars that tell the app which art assets and config values to use
	* Javascript is run through Babel with es2015, stage-0, and react presents, as well as transforms to render decorators, which is needed for a third-party plugin. here the build constant symbols are replaced with strings.
	* the constants are also injected into SCSS, and the result is interpreted with a chain of loaders: constants -> sass-loader -> css-loader -> style-loader. The css-loader resolves most of the constants into urls and then into file references or data URIs, as appropriate. Ignored in Wallaby.
	* the result of the above is added as a `style` attribute to the JSX components which make up the views of the app. Ignored in Wallaby.
	* the whole thing is bundled up with dependencies (important) and made browser-friendly (less important)

So Webpack runs in a node environment, as is necessary, and the tests are written using fgnass/domino, so they are also able to run in node even though they often depend on browser activity (eg. testing iframe messaging.) I don't understand why I can run PhantomJS2 with karma and have the tests execute, but wallaby has trouble with the same setup. Although wallaby itself reliably executes the config file, it doesn't seem to run Webpack before Babel, as it uses node's module.js to compile and execute the `files` argument.

I would be able to cut out webpack completely, but I need the build constant plugin - and really, I don't want to miss any potential problems that it might introduce. For example, using the compiler option in wallaby alerted me to the fact that the order you define presets in is important - having stage-0 before es2015 really made a mess until I figured that out, even though it seems to matter less when using Webpack loaders.

So, that's my overall usecase. Any ideas on how I can defer executing the javascript until after Webpack has injected the constants? I'm attaching my config below (again). Like you saw yesterday, I've tried to have it served in the browser & tests run against that, but ran into other problems (namely the inability to `require` things - Webpack doesn't run first then either, it seems.)

http://pastebin.com/wbhXq4Wy

**Comment by ArtemGovorov** [comment]
> I don't understand why I can run PhantomJS2 with karma and have the tests execute, but wallaby has trouble with the same setup.

It doesn't look like it's the same setup. In your wallaby config you are setting these:
``` JavaScript
return {
    env: {
      ...
      type: 'node',
      runner: 'node'
```

which makes wallaby **run your tests in node.js** (while you have karma running your tests in PhantomJs2). You need to remove these:
``` JavaScript
      type: 'node',
      runner: 'node'
```
and uncomment the line where you're setting PhantomJs2 as a `runner`.

> Webpack runs in a node environment

Yes, and wallaby will run it for you in node environment (as well as karma will do), but you don't have to configure `env` for that. As [described in the docs section](http://wallabyjs.com/docs/config/runner.html) that I have referenced in the previous issue, `env` is to describe where to run your tests and not how to compile/prepare your code with webpack. `env` in wallaby is like `browsers`/`customLaunchers` in karma. As for the `NODE_ENV` and `BABEL_ENV` that are required for webpack, you may set them right in your config (there's a sample in the same docs section):
``` JavaScript
process.env.BABEL_ENV = 'test';
```

You also need to uncomment the line `window.__moduleBundler.loadTests();` in your bootstrap function.

Just to sum up two previous points: both karma/wallaby run its config files using node.js, those config files invoke webpack to build code, once everything is ready, karma/wallaby run tests in a configured runner/environment. For running tests, karma only supports browsers and Phantom, wallaby supports Phantom and node.js (for node.js projects). What your config is trying to do is to run `webpack`-ed tests in node.js with wallaby and this scenario is not supported (neither by wallaby, nor by karma that doesn't support running tests in node.js at all). Hope it makes sense.

Regarding another issue:

>  it doesn't seem to run Webpack before Babel,

Correct, wallaby.js runs its Babel compiler before Webpack. In your specific case, if I understood correctly, you need to run Babel compilation after certain other webpack transformations (because the third-party babel plugin is using the result of DefinePlugin work?). 

In this case, you may just use Webpack loaders for babel that you have and remove wallaby babel compilers section all together. As [mentioned in the docs section](http://wallabyjs.com/docs/integration/webpack.html#loaders-vs-preprocessors-compilers), wallaby compiler for babel instead of webpack loader is an optional thing. So just add your babel loader(s) to your webpack config in wallaby.js file and try that.

**Comment by karlnp** [comment]
Ah, ok. That shouldn't be an issue then. Thanks very much for the clarification! I was hoping to take advantage of the faster Wallaby compiler, but until we have a better solution for our constants I may just have to deal with it. 

I'm not at all sure whether the problem lies with babel's execution time - when I run tests in the browser, I get 'cannot find require' from my test file, which should be webpacked - ah, but I've just noticed that it's set to load the test in a script tag instead of being wrapped by webpack. 

Well, I'll give it a shot on Monday and see if that makes things work! Thanks again.

**Comment by ArtemGovorov** [comment]
> I was hoping to take advantage of the faster Wallaby compiler

You may still try doing it, but I suggest to get the whole thing working first before optimising it. Wallaby `compiler` is [just a function](http://wallabyjs.com/docs/config/compilers.html) that takes a source code as an input and expects a transformed code (+ a couple of other things) as an output, so you may define a function as a wallaby compiler that would call `DefinePlugin` (or whatever it uses internally) and then would call wallaby babel compiler.

> but I've just noticed that it's set to load the test in a script tag instead of being wrapped by webpack

Nice spot, it should certainly be `load: false` for the webpack way.

**Comment by karlnp** [comment]
Okay, I've re-added the loaders and it's working the best it ever has - but I still have inscrutable problems. Namely:

```
ModuleParseError: Module parse failed: .\app\game-wrappers\gamesInterface.js Line 9: Unexpected token
You may need an appropriate loader to handle this file type.
| var Actions = ($_$w(26, 5), Constants.ActionTypes);
| $_$w(26, 6), require('string_score');
| export default class {
|     buildAction(suppliedType, suppliedJson) {
|         $_$wf(26);
```

It seems like it might be talking about `export default`, but I have the es2015 preset active, and I only get this error on the module that my test imports. It is strange that babel is interpreting files that have already been interpreted through webpack, though - might this have something to do with `wallaby-webpack`?

**Comment by karlnp** [comment]
It looks like the babel-loader isn't triggering? If I add an `import` statement to my test it fails there at the same place in `wallaby-webpack`: `WallabyInputFileSystem:30`, and then down the Webpack chain to NormalModuleMixin/NormalModule.js.

```
ModuleParseError: Module parse failed: .\app\game-wrappers\gamesInterface-Test.js Line 3: Unexpected token
You may need an appropriate loader to handle this file type.
| $_$wp(27);
| var domino = ($_$w(27, 0), require('domino'));
| import gamesInterface from './gamesInterface';
| $_$w(27, 1), describe('gamesInterface', function () {
|     $_$wf(27);
  at DependenciesBlock.<anonymous> (C:\Users\Karl\git\gtp\node_modules\webpack\lib\NormalModule.js:113:20)
  at DependenciesBlock.onModuleBuild (C:\Users\Karl\git\gtp\node_modules\webpack\node_modules\webpack-core\lib\NormalModuleMixin.js:310:10)
  at nextLoader (C:\Users\Karl\git\gtp\node_modules\webpack\node_modules\webpack-core\lib\NormalModuleMixin.js:275:25)
  at C:\Users\Karl\git\gtp\node_modules\webpack\node_modules\webpack-core\lib\NormalModuleMixin.js:259:5
  at C:\Users\Karl\git\gtp\node_modules\wallaby-webpack\lib\WallabyInputFileSystem.js:30:11
  at doNTCallback0 (node.js:419:9)
  at process._tickCallback (node.js:348:13)
```

**Comment by karlnp** [comment]
As far as I can tell using Rewire messes up webpack and/or babel, regardless of whether I use `rewire-webpack` or `babel-plugin-rewire`. That's frustrating.

**Comment by ArtemGovorov** [comment]
> It looks like the babel-loader isn't triggering?

Yes, it looks like webpack babel loader is not triggering. Make sure the loader `test` property is correct and includes your files. 

> As far as I can tell using Rewire messes up webpack and/or babel

Do you mean it's causing the previous issues or you see some new issues caused by using Rewire? I'm a bit lost as I don't remember seeing rewire in any config file that you've shared.

I think the best way would be if you could create a small sample demonstrating all issues, perhaps just by  taking your code and taking out everything business sensitive, changing folder names, leaving one test and minimal code, and publishing it in a GitHub repo (or emailing me). This way we'll be able to quickly sort out all issues.

**Comment by ArtemGovorov** [comment]
I'm pretty sure we are close to make it working and with the sample that I could try and run myself to see all the issues, we'll make it work.

**Comment by karlnp** [comment]
Rewire (or a clone) is used on tests in the above linked config/pastebin using `babel-plugin-rewire`, which is actually a reimplementation of rewire rather than an adaptor for the base package. But regardless of whether I use that or the base package with `rewire-webpack`, something in this hideous mess of a toolchain balks at that.

I actually went back and refactored some code to work with passed-in functions as callbacks and used sinon instead so I could get back to work, but [that's not much better for wallaby, apparently](http://puu.sh/mAU7r/26f11b185b.png). I'll try and make a sample repo this week, but I have some productivity to catch up to in the meantime - and making a suitably complex repo will require some dedication.

I still really appreciate your help, thank you!

**Comment by ArtemGovorov** [comment]
Awesome, thanks! Sorry for being help faster and more efficiently, with non-trivial setups like your so many things can be misconfigured and it's hard to see what's wrong without trying it.

In a meantime, could you please share your latest wallaby config? I could try re-constructing your structure and creating the sample myself so you could jump in later and we could start collaborating on it.

**Comment by karlnp** [comment]
Sure, I'd be happy to collaborate!

[Here you go.](http://pastebin.com/zwjPWAzc) Some things are extraneous, like the babel polyfill - I was trying a lot of things in debugging so it's kind of a mess.

Like I said above, the weirdest case is actually undefined variables that have injected definitions at webpack compile time, but that doesn't actually seem to be a problem in practice. The unit under test is a simple wrapper for browser events, with a virtual DOM in use so that it can be tested in node as well should it end up being required. 

I totally understand what you mean about the difficulty of testing non-trivial setups. Particularly when there are so many moving parts and transpiling/concatenating/packaging modules in play.

**Comment by ArtemGovorov** [comment]
Hmm, I thought you were going to remove wallaby compilers section all together and replace it with abel-loader? But in the config I still wallaby compilers and no babel-loader in webpack config.

**Comment by ArtemGovorov** [comment]
Also could you please share your working karma config and your webpack config that karma is using?

**Comment by karlnp** [comment]
Ah, right. Here's the loaders section - I've just been switching back and forth as necessary to test different things, and went back to `compilers` approach as a way to 'simplify' things, I suppose.

with devtool: 'cheap-module-source-map':
```
loaders: [{
        test: /\.jsx?$/,
        exclude: /(node_modules|bower_components|vendor)/,
        loader: 'babel',
        query: {
          sourceMaps: true,
          presets: ['es2015', 'stage-0', 'react'],
          plugins: ['transform-runtime', "babel-plugin-transform-decorators-legacy"]
        },
        include: [appRootPath,
          path.join(appRootPath, "jsx-templates"),
          path.join(appRootPath, "sass"),
          path.join(appRootPath, "dispatchers"),
          path.join(appRootPath, "stores")]
      },
        {
          test: /\.js?$/,
          exclude: /(node_modules|bower_components|vendor)/,
          loader: 'babel',
          query: {
            sourceMaps: true,
            presets: ['es2015', 'stage-0', 'react'],
            plugins: ['transform-runtime', "babel-plugin-transform-decorators-legacy"]
          },
          include: [
            appRootPath,
            path.join(appRootPath, "views"),
            path.join(appRootPath, "server_interfaces"),
            path.join(appRootPath, "dispatchers"),
            path.join(appRootPath, "game-wrappers")
          ]
        }
      ],
```

Here's the karma config - as I note, the webpack config is essentially the same as in wallaby.js. I've been trying to converge on a set of configs that all run off the same webpack config and override as needed to enable continuous integration.

http://pastebin.com/AyyEKLv2

**Comment by ArtemGovorov** [comment]
Thanks, I'll try to put the sample together and publish it, hopefully will be able to reproduce those issues.

**Comment by karlnp** [comment]
Potentially useful things I tried and failed to do:

* added a babel preprocessor, removed babel-loaders - running the test fails with "Cannot find module: 44", I assume because wallaby-webpack has already been and gone
*  added babel-loaders back, changed the test entry in the files list to 'load: false', everything successfully compiles but 0 tests are detected/run

I'm about out of ideas for today. Good luck on your end!

**Comment by ArtemGovorov** [comment]
So I have created this simple repo - https://github.com/ArtemGovorov/wallaby-402, where I have reproduced some of your structure, added a simple test and one dependency. Everything works fine:

<img width="406" alt="screen shot 2016-01-19 at 12 00 10 pm" src="https://cloud.githubusercontent.com/assets/979966/12407619/889d5f80-bea4-11e5-84ba-7406261e760b.png">

Could you please try adding your project specific things to it to try to break it?

**Comment by ArtemGovorov** [comment]
After you clone the repo, you'll need to run `npm install` and also replace the runner path with your phantomjs exe if you're on Window (can be from the same module or from your local folder)
```
    env: {
      runner: "node_modules/phantomjs2-ext/lib/phantom/bin/phantomjs"
    },
```

**Comment by karlnp** [comment]
https://github.com/ArtemGovorov/wallaby-402/pull/1

**Comment by karlnp** [comment]
The problem is definitely with how gamesInterface.js is exported. I thought at first it was with duplicate property names between the exported functions and the default object, but deleting the default export entirely doesn't seem to make a difference. My stack trace is 2 lines long and doesn't make sense:
```
ReferenceError: Can't find variable: require
	at app/game-wrappers/gamesInterface.js:24
	at app/game-wrappers/gamesInterface-Test.js:3

```
so I'm hoping you have more visibility into the source with the actual source at hand. It's also possibly related to https://phabricator.babeljs.io/T2212, but I don't think it is at first glance?

**Comment by karlnp** [comment]
May also be related to https://phabricator.babeljs.io/T6748.

**Comment by karlnp** [comment]
I think ES6 modules, regardless of how they are defined, are probably unable to be mocked in the same way that frozen objects are unable to be mocked. On the other hand, because Babel treats all modules as ES6 modules, any exports are reassigned to exports.default and therefore decontextualized, so we can't just get around it by writing CommonJS modules with ES6 features. Also, increasingly the dominant format for module documentation is in ES6 (as well as, for example, React being written more and more as an ES6 lib), so abandoning it for ES5 is impractical.

Hmm.

**Comment by ArtemGovorov** [comment]
@karlnp Thanks for the pull request! 

We'll get to the mocking story, let's do one step at a time and first try to sort out the issue with `ReferenceError: Can't find variable: require`. 

So I'm investigating it and it looks like the issue is not with export/import, but (surprise) with `JSON.stringify` call (or the way babel/webpack process it). I'll keep digging, you may have a look at my latest commit (with the https://github.com/ArtemGovorov/wallaby-402/blob/master/app/game-wrappers/gamesInterface.js#L84 line commented out) - it starts working. Well, it throws a different error, but that error sounds reasonable and is related to `keyMirror` usage.

**Comment by karlnp** [comment]
Yeah, it works fine as long as I don't use JSON.stringify in that unused function. What the hell? 

I mean, I guess it's a workaround, but being as this app is a REST API wrapper, we use that quite a lot...

**Comment by ArtemGovorov** [comment]
Here's where what happens - there's the `'transform-runtime',` plugin used (it also starts working if you remove from the sample config). When the plugin is used, it replaces `JSON.stringify` call with something like `require("core-js/library/fn/json/stringify")()` and because we don't have the core-js module installed in the sample, webpack (silently) fails to resolve the dependency and just leaves the code as is, so when running it we get the `Can't find variable: require`. I'll now try to install core-js to see if it fixes the issue.

**Comment by karlnp** [comment]
core-js is part of babel-core, though. At least, it should be.

**Comment by ArtemGovorov** [comment]
That's what I thought, but nevertheless, babel generates `require("core-js/library/fn/json/stringify")` and webpack will try to search for the `core-js` module only in the root `node_modules` (unless there's some `resolve.fallback` option is set).

Installing core-js didn't help, I'll keep investigating further. Wondering how does it actually work in your karma setup. Are you using the same webpack and babel versions there? Perhaps you could add the working karma config to the sample? That'd help me a lot in my investigation.

**Comment by karlnp** [comment]
Sure. It hasn't really changed, though. Although I am having a hard time getting it to repro with a minimum config file, even copied directly across...

**Comment by ArtemGovorov** [comment]
@karlnp I think we're getting somewhere. If I comment out this line in the webpack config - https://github.com/ArtemGovorov/wallaby-402/blob/master/wallaby.js#L20, it starts working.

**Comment by ArtemGovorov** [comment]
So it looks like the generated module (that contains `require("core-js/library/fn/json/stringify")`) is not parsed by webpack. That's why it doesn't throw any errors - it just leaves the code as is. But the code contains `require("...")` so has to be `webpack`-ed.

**Comment by ArtemGovorov** [comment]
Ok, the mystery is solved. To restore the whole picture:
- babel compiles your `gamesInterface.js` file and `JSON.stringify` in it gets compiled to something like `require('babel-runtime/core-js/json/stringify')()`. BTW, you may see the compiled files (both `babel`-ed and `webpack`-ed) in the [wallaby cache](http://wallabyjs.com/docs/intro/troubleshooting.html#instrumented-code-and-generated-test-page) folder.
- when webpack compiles your `gamesInterface.js` after babel (the same if you use babel-loader), it sees the `require('babel-runtime/core-js/json/stringify')()` and resolves the `babel-runtime/core-js/json/stringify` module to compile it as well.
- the module has `json` in its path, so your `noParse: /(scss|react\.js|sinon\/).*/` regexp tells webpack that it should NOT parse the module, but insert its content in the bundle as is.
- the `babel-runtime/core-js/json/stringify` content is `module.exports = { "default": require("core-js/library/fn/json/stringify"), __esModule: true };`, so the `require(...)` code ends up in the bundle and of course the browser fails to execute it.

BTW, if you're using exactly the same webpack config in karma, it should also fail (I don't see any reasons why it wouldn't).

**Comment by karlnp** [comment]
https://github.com/ArtemGovorov/wallaby-402/pull/2

Interestingly, it's not exactly the same - it has a few minor differences, most significant of which is a different sourcemap source. It does still have the regex, so maybe we're seeing two different resolution methods.

**Comment by karlnp** [comment]
It's not the whole story, because a corrected regex results in the error 
```
Error: Cannot find module '60'
	at app/game-wrappers/gamesInterface.js:24
	at app/game-wrappers/gamesInterface-Test.js:7
```

and indeed there is no module 60 in the cache folder. The structure.json only goes up to 46, but I could easily not be reading that properly.

**Comment by ArtemGovorov** [comment]
@karlnp I may investigate why does it work in karma, perhaps some other setting is shadowing the `noParse`.

I have just pushed the working version (removed `json` from the RegExp and have removed babel-loaders from wallaby config, you either need those or wallaby compilers, not both) . Try to see of it works for you (also change your phantom path).

**Comment by karlnp** [comment]
We're back in business on the sample repo, but this is what I get on the original app, despite carrying over the changes from your working version:

```
Error: No source available
	at app/game-wrappers/gamesInterface-Test.js:1

ModuleParseError: Module parse failed: .\app\game-wrappers\gamesInterface-Test.js Line 2: Unexpected token
You may need an appropriate loader to handle this file type.
| $_$wp(30);
| import * as domino from 'domino';
| import * as gamesInterface from './gamesInterface.js';
| import * as util from 'util';
  at DependenciesBlock.<anonymous> (C:\Users\Karl\git\gtp\node_modules\webpack\lib\NormalModule.js:113:20)
  at DependenciesBlock.onModuleBuild (C:\Users\Karl\git\gtp\node_modules\webpack\node_modules\webpack-core\lib\NormalModuleMixin.js:310:10)
  at nextLoader (C:\Users\Karl\git\gtp\node_modules\webpack\node_modules\webpack-core\lib\NormalModuleMixin.js:275:25)
  at C:\Users\Karl\git\gtp\node_modules\webpack\node_modules\webpack-core\lib\NormalModuleMixin.js:259:5
  at C:\Users\Karl\git\gtp\node_modules\wallaby-webpack\lib\WallabyInputFileSystem.js:30:11
  at doNTCallback0 (node.js:419:9)
  at process._tickCallback (node.js:348:13)

```

What a convoluted mystery this is!

**Comment by ArtemGovorov** [comment]
The error is from webpack, it for some reason ends up receiving ES6 code to bundle. It means that for some reason, wallaby compilers didn't run on the file. Check your patterns for the file, I remember you had 3 different compiles defined, try to just do a single compiler with the `**/*.js*` pattern.

**Comment by karlnp** [comment]
Do they run on tests as well, or just the files entry in wallaby.js?

**Comment by ArtemGovorov** [comment]
They run on all files listed in `tests` and `files`.

**Comment by ArtemGovorov** [comment]
So for example, if you have:
```
files: ['src/**/*.js'],
tests: ['test/**/*.js']
compilers: {
  '**/*.js': ...
}
```

then the compiler will run on all JS files in `src` and `test` folders.

**Comment by ArtemGovorov** [comment]
* but not on `node_modules` that are not in any of the lists (and shouldn't be).

**Comment by karlnp** [comment]
`'**/*.{react.}js*'` should greedy-match everything, including *.react.js files, right? For some reason it continues to tell me that import * is an unexpected token. Or maybe it's talking about  `$_$wp(30);`? 

I'll think about this on the way home & try to get my computer to run Webstorm.

**Comment by karlnp** [comment]
I continue to appreciate your support on this.

**Comment by ArtemGovorov** [comment]
It's definitely telling about `import *`. Webpack can't compile and bundle ES6, something has to compile it to ES5 before it tries to bundle it. In prod config it would be babel-loader, in our case, wallaby babel compiler is supposed to do to, but it doesn't. The only reason I may suggest is that the file path doesn't satisfy the compiler pattern, so I suggest trying a single simple pattern: `'**/*.js'`, does it work for you? I'm not not sure how the pattern with `react.` is related, the file that fails is `.\app\game-wrappers\gamesInterface-Test.js`, it doesn't have `react.` in it.

**Comment by karlnp** [comment]
Unrelated files with the extension .react.jsx  (that need to be transpiled) fail when I use that simple extension, as well as the test. When I use the above pattern only the test fails to load in webpack.

**Comment by ArtemGovorov** [comment]
Sorry, I meant `'**/*.js*'`, not just `'**/*.js'`. Perhaps they fail because `noParse` RegExp contains `react` in it? 

Anyway, let's do one step at a time and try to find out why the test file fails to load in webpack. It definitely satisfies both `'**/*.js'`/`'**/*.js*'` patterns, so something else it going on, but without a way to reproduce I have no idea what. Now given that we have a working sample, could you please try breaking it by adding things that are different between your wallaby config the sample wallaby config?

**Comment by karlnp** [comment]
I will, but I won't be able to do so tonight. (And the noParse only has vendor & bower_components in it at this point.) Your revised one is what I expected you meant & is the one that did not match, probably because the . is greedily matched and so the first . of ".react.jsx" is aggressively excluded. I don't have any visibility into how wallaby interprets the file glob, though.

**Comment by ArtemGovorov** [comment]
> I don't have any visibility into how wallaby interprets the file glob, though.

Exact same way as the most of node modules (like karma, grunt, gulp, etc.) do, we use [minimatch](https://github.com/isaacs/minimatch).

**Comment by ArtemGovorov** [comment]
You may also find more samples here - https://github.com/isaacs/node-glob#glob-primer

**Comment by karlnp** [comment]
Today's theory: Since the glob (and the refined version: `'**/*?(.react)?(-Test).js*'`) actually work, as shown by wallaby's compile output log, the problem is somewhere downstream causing babel to fail silently.. I suspect, in my case, it's the symbols which are defined at webpack time and which would be undefined for babel.

So, for lack of a better idea, I'm setting them on the node global var at wallaby.js file parse time and trying to ensure that keyMirror doesn't mess things up by using a FP function that should do the same thing. If that doesn't work, I'll define them directly as a fallback.. but it might mess up my build process. We'll see.

**Comment by karlnp** [comment]
Think I got it - it's an uncaught exception in the sandbox from over-stripped strings. Adding the case to your repo shortly.

**Comment by karlnp** [comment]
https://github.com/ArtemGovorov/wallaby-402/pull/3

**Comment by karlnp** [comment]
Ha! Got everything working. The remaining unclear bug was caused by a circular dependency.. fortunately an *upside* of es2015 modules is that they don't cause circular dependencies, so once I converted the central module everything went smoothly.

This was a fun week, huh?

**Comment by ArtemGovorov** [comment]
@karlnp Merged and reproduced it, thanks. From what I can see, there are 2 separate issues:
- wallaby is not reporting the issue correctly (reporting only a part of the issue, and not the helpful part with syntax error),
- invalid build constants file, string values are supposed to be defined as described here: https://webpack.github.io/docs/list-of-plugins.html#defineplugin, maybe with JSON.stringify(...), otherwise webpack just substitutes your `WebpackAPIRoot` with `https://url2.net` and that breaks the file syntax. I have pushed the changes to the `build_constants.js` file.

Re the first issue, I'll be looking into it and fixing, but it's not blocking.
Re the second issue, is it a real issue in your production code or is it just to demonstrate the first issue? The `build_constants.js` is not valid for webpack substitutions in it's current state (for strings).

**Comment by ArtemGovorov** [comment]
> Got everything working.

Awesome! Thanks for the update. Glad it finally works. I'll be looking into the insufficient reporting issue that you've helped to reproduce and hopefully fixing it shortly.

**Comment by karlnp** [comment]
Yep! I suspect the insufficient reporting might be fixed by a more generous try/catch, since it's really the uncaught exception crashing phantomjs' javascript VM that makes it so hard to track down. For what it's worth, my circular dependencies had a similar issue, where the report was
`TypeError: undefined is not a constructor (resolving module.method)` and looking at the VM output helped tremendously there as well.

The second issue isn't in our production code, but we did run into the same thing. I'd just forgotten to move the change over, but it turned out to be a productive mistake, at least.

**Comment by ArtemGovorov** [comment]
Thanks a lot for your help! I have a suspicion that the insufficient reporting may be a PhantomJs thing, we do subscribe to all of its errors, but the syntax one is not being reported (my suspicion with that PhantomJs in karma will report the error the same way unfortunately). Chrome does things better than PhantomJs, so that's why we can see the error in DevTools when opening the sandbox page.

**Comment by karlnp** [comment]
I'm still flying pretty high from successfully building my complete app w/ tests, so I might poke at it tomorrow, since I'm pretty handy with reflection & introspection now. Worth a shot. Thank you as well!

**Comment by ArtemGovorov** [comment]
@karlnp Thanks, if you hit any more bumps, feel free to update the sample repo, happy to jump in and have a look.

**Comment by karlnp** [comment]
I hit some gnarly stack traces in wallaby/server.js at points today - I'll probably repro them tomorrow to help you out.

**Comment by ArtemGovorov** [comment]
@karlnp I have fixed the reporting issue and published the new core version (should be updated automatically when you start wallaby.js). Now if you try providing an invalid `build_constants.js` it should point out much better where the issue is.
